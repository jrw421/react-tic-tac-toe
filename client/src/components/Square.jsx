import React, { Component } from 'react';

export default class Square extends Component {
    constructor(props) {
        super(props)
    }

    render() {
        let { value, position, takeTurn } = this.props;
        return (
            <div onClick={() => takeTurn(position)} className="square">
                <div className="value">{value}</div>
                {/* <div>{position}</div> */}
            </div>
        )
    }
}