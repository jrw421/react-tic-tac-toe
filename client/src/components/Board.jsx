import React, { Component } from 'react';
import Square from './Square.jsx';

export default class Board extends Component {
    constructor(props) {
        super(props);
        this.state = {
            squares: Array(3).fill([null, null, null]),
            isXturn: true
        }
        this.takeTurn = this.takeTurn.bind(this);
        this.checkForWin = this.checkForWin.bind(this);
    }

    takeTurn(positions) {
        let positionX = positions[0];
        let positionY = positions[1];
        let { isXturn } = this.state;
        let move = isXturn ? 'X' : 'O';

        let updatedSquares = this.state.squares.map(row => row.map(val => { return val }));
        updatedSquares[positionX][positionY] = move;

        this.setState({
            squares: updatedSquares,
            isXturn: !this.state.isXturn
        }, () => this.checkForWin(positionX, positionY));
    }   

    checkForWin(positionX, positionY) {
        let { incrementWin } = this.props;
        let winHorizontally0 = (this.state.squares[0][0] !== null) && this.state.squares[0][0] === this.state.squares[0][1] &&  this.state.squares[0][1] === this.state.squares[0][2];
        let winHorizontally1 = (this.state.squares[1][0] !== null) && this.state.squares[1][0] === this.state.squares[1][1] &&  this.state.squares[1][1] === this.state.squares[1][2];
        let winHorizontally2 = (this.state.squares[2][0] !== null) && this.state.squares[2][0] === this.state.squares[2][1] &&  this.state.squares[2][1] === this.state.squares[2][2];

        let winVertically0 = (this.state.squares[0][0] !== null) && this.state.squares[0][0] === this.state.squares[1][0] &&  this.state.squares[1][0] === this.state.squares[2][0];
        let winVertically1 = (this.state.squares[0][1] !== null) && this.state.squares[0][1] === this.state.squares[1][1] &&  this.state.squares[1][1] === this.state.squares[2][1];
        let winVertically2 = (this.state.squares[0][2] !== null) && this.state.squares[0][2] === this.state.squares[1][2] &&  this.state.squares[1][2] === this.state.squares[2][2];

        let winDiagonally0 = (this.state.squares[0][0] !== null) && this.state.squares[0][0] === this.state.squares[1][1] &&  this.state.squares[1][1] === this.state.squares[2][2];
        let winDiagonally1 = (this.state.squares[0][2] !== null) && this.state.squares[0][2] === this.state.squares[1][1] &&  this.state.squares[1][1] === this.state.squares[2][0];
        
        const horizontalWin = winHorizontally1 || winHorizontally2 || winHorizontally0;
        const verticalWin = winVertically1 || winVertically2 || winVertically0;
        const diagonalWin = winDiagonally0 || winDiagonally1;

        if (horizontalWin || verticalWin || diagonalWin) {
            let winner = this.state.squares[positionX][positionY];
            alert(`${winner}'s won!`)
            incrementWin(winner);
            this.setState({
                squares: Array(3).fill([null, null, null])
            })
        }
    }

    render() {
        return (
            <div className="grid">
                {this.state.squares.flatMap((row, i) => {
                    return row.map((value, j) => { return <Square value={value} key={Math.random()} takeTurn={this.takeTurn} position={[i, j]}/> }
                )})}
            </div>
        )
    }
}