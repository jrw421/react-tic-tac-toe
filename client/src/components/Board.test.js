import Board from './Board.jsx';
import Square from './Square.jsx';
import React from 'react';
import renderer from 'react-test-renderer';

const state = {
    horizontalWin: [['X','X','X'], [null, null, null], [null, null, null]],
    verticalWin: [['X', null, null], ['X', null, null], ['X', null, null]],
    diagonalWin: [['X', null, null], [null, 'X', null], [null, null, 'X']]
}

const board = renderer.create(<Board/>).getInstance();

test('Board solution to horizontally win', () => {
    expect(board.checkForWin(0, 0, state.horizontalWin)).toBe(true);
});

test('Board solution to vertically win', () => {
    expect(board.checkForWin(0, 0, state.verticalWin)).toBe(true);
});

test('Board solution to diagonally win', () => {
    expect(board.checkForWin(0, 0, state.diagonalWin)).toBe(true);
});

describe('Board component should render', () => {
    it('renders correctly', () => {
        const rendered = renderer.create(
        <Board/>
        );
        expect(rendered.toJSON()).toMatchSnapshot();
    });
});