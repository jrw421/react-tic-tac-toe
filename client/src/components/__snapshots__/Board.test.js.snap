// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component should render renders correctly 1`] = `
<div
  className="grid"
>
  <div
    className="square"
    onClick={[Function]}
  >
    <div
      className="value"
    />
  </div>
  <div
    className="square"
    onClick={[Function]}
  >
    <div
      className="value"
    />
  </div>
  <div
    className="square"
    onClick={[Function]}
  >
    <div
      className="value"
    />
  </div>
  <div
    className="square"
    onClick={[Function]}
  >
    <div
      className="value"
    />
  </div>
  <div
    className="square"
    onClick={[Function]}
  >
    <div
      className="value"
    />
  </div>
  <div
    className="square"
    onClick={[Function]}
  >
    <div
      className="value"
    />
  </div>
  <div
    className="square"
    onClick={[Function]}
  >
    <div
      className="value"
    />
  </div>
  <div
    className="square"
    onClick={[Function]}
  >
    <div
      className="value"
    />
  </div>
  <div
    className="square"
    onClick={[Function]}
  >
    <div
      className="value"
    />
  </div>
</div>
`;

exports[`Square component renders the move correctly renders correctly 1`] = `
<div
  className="square"
  onClick={[Function]}
>
  <div
    className="value"
  >
    X
  </div>
</div>
`;
